### DISCUSSION TITLE
Access-Control-Allow-Methods does not work

### DISCUSSION THREAD

**Discussion by mac-ro-ni** [dotnet/WatsonWebserver#99] [General]
I'm try to set Access-Control-Allow-Methods with a custom http request method types but it does not show on actual response headers

Custom Method : "RDSERVICE"

`WebServer.Settings.Headers["Access-Control-Allow-Methods"] = "RDSERVICE";`

**Comment by mac-ro-ni** [comment]
I was setting the request header.
Please guide me to set response header.

**Comment by mac-ro-ni** [comment]
**Update**
I have added Custom method to response headers in DefaultRoute using 
`ctx.Response.Headers["Access-Control-Allow-Methods"] = "RDSERVICE";`
It is working fine with GET Request but when i try sending OPTIONS request it shows default methods i.e `OPTIONS, HEAD , GET, PUT, POST, DELETE`

**Comment by jchristn** [comment]
Did you try overloading ```Routes.Preflight```?

**Comment by mac-ro-ni** [comment]
I didn't try that. please give an example for that

**Comment by jchristn** [comment]
You can see the default implementation here: https://github.com/jchristn/WatsonWebserver/blob/653fd49292faefe009a29acdcaaae81d715d69de/WatsonWebserver/WatsonWebserverRoutes.cs#L185

**Reply by mac-ro-ni** [reply]
@jchristn  Thank you so much!

**Comment by mac-ro-ni** [comment]
After setting preflight cors error is resolved but the response is not being sent from the request

**Reply by jchristn** [reply]
Could you show me your code?

**Comment by mac-ro-ni** [comment]
WebServer:
``` 
WebServer = new Server("127.0.0.1", 11100, false, DefaultRoute); 
WebServer.Routes.Preflight = PreflightRoute;
WebServer.Start();
```

Routes:
``` 
 private async Task DefaultRoute(HttpContext arg)
        {
            String sourceIp = arg.Request.Url.Full;
           // HTTPIncomingRequest.Invoke(new Action(delegate () { HTTPIncomingRequest.Text = sourceIp; }));
            //Logcat.Invoke(new Action(delegate () { Logcat.Text += arg.ToJson(); }));
            arg.Response.Headers["Access-Control-Allow-Methods"] = "RDSERVICE";
            await arg.Response.Send("Hola");

        }

 private async Task PreflightRoute(HttpContext arg)
        {
            String sourceIp = arg.Request.Source.IpAddress;
           // HTTPIncomingRequest.Invoke(new Action(delegate () { HTTPIncomingRequest.Text = sourceIp; }));
          //  Logcat.Invoke(new Action(delegate () { Logcat.Text += "Preflight" + arg.ToJson(); }));
            arg.Response.Headers["Access-Control-Allow-Methods"] = "RDSERVICE";
            arg.Response.Headers["Content-type"] = "application/xml";
            await arg.Response.Send().ConfigureAwait(false);

        }

```
Developer Tools  Screenshot: https://i.imgur.com/vyDH6El.png

**Reply by jchristn** [reply]
What happens if you set the response content length to zero prior to calling ```.Send()```?

**Reply by mac-ro-ni** [reply]
It still shows pending.  I forgot to mention that the request method is `RDSERVICE`. Can it handle these custom requests?

**Reply by mac-ro-ni** [reply]
The request stays on pending until i stop the server

**Reply by jchristn** [reply]
Got it.  It's probably throwing an exception because it expects the method to be one of the set in the enum.

https://github.com/jchristn/WatsonWebserver/blob/653fd49292faefe009a29acdcaaae81d715d69de/WatsonWebserver/HttpRequest.cs#L203

**Reply by jchristn** [reply]
And also: https://github.com/jchristn/WatsonWebserver/blob/master/WatsonWebserver/HttpMethod.cs

**Reply by jchristn** [reply]
Net-net, today it doesn't support HTTP methods other than those in the enum.  If you wrap your code in try/catch, I'm pretty sure you'll see an exception on that line.

**Reply by mac-ro-ni** [reply]
@jchristn  is there any way i can override these enums and add custom method.

**Reply by mac-ro-ni** [reply]
I tried something like this but it still doesn't work
```
HttpMethod RDService;
Enum.TryParse("RDSERVICE", out RDService);
WebServer.Routes.Static.Add(RDService, "/", RDRoute);
```

**Reply by jchristn** [reply]
You could always clone the library and add whatever you wanted to the enum.  Alternatively I could amend the library to not throw if the request method is unknown, but rather set it to a value like ```Unknown``` and then include another property in the ```HttpRequest``` (as a string) that you could then evaluate on your side.  Would that work?

**Reply by mac-ro-ni** [reply]
@jchristn Adding `Unknown` would be great. I hope it will work

**Reply by jchristn** [reply]
Which framework are you using?

**Reply by mac-ro-ni** [reply]
.NET Framework

**Reply by jchristn** [reply]
Version?

**Reply by mac-ro-ni** [reply]
4.6.2

**Reply by jchristn** [reply]
Can you give v4.3.0 a try and let me know?  ```Method``` will be ```UNKNOWN``` and ```MethodRaw``` will contain the string value (in ```HttpRequest```)?

NuGet: https://www.nuget.org/packages/Watson/4.3.0
Commit: https://github.com/jchristn/WatsonWebserver/commit/c73685143b283e98a4cd807156c18994d818251c

Cheers, Joel

**Reply by mac-ro-ni** [reply]
wow, it works like a charm!!
Thank you so much @jchristn

**Reply by mac-ro-ni** [reply]
One more question @jchristn. I want to get some data from the file and send the response. How to delay the response until i read the files ? Maybe `CountownEvent`?

**Reply by jchristn** [reply]
Have some sample code I could take a look at?

**Reply by mac-ro-ni** [reply]
Something like this ! Getting data from API and using that response data to send it using Watsonserver. 
```
CountdownEvent countEventObject = new CountdownEvent(1);
String rdResponse  = null;
private async Task anotherServerResponse(String res)
        {
           // Request Recieved From another server
           rdResponse = res;
          countObject.Signal();
        }

private async Task rdServiceRoute(HttpContext arg)
        {
           // Request Sent to another Server
            anotherServerRequest();
            countEventObject.Wait();
            await arg.Response.Send(rdResponse);

        }

```

**Reply by jchristn** [reply]
Yeah, I would either use a synchronous call or await an asynchronous call to the other server, i.e.
```
private async Task DefaultRoute(HttpContext ctx)
{
  await CallTheOtherServer();
  await ctx.Response.Send();
}

private async Task CallTheOtherServer()
{
  // call the other server here
}
```
You could also do what you are suggesting using ```Wait()```

**Reply by mac-ro-ni** [reply]
Can you give me an example code for getting Response from TCP Server and serving it as a response.

**Reply by jchristn** [reply]
It's going to depend on how you implement your TCP client library. I have a few out there (SuperSimpleTcp being one) and there are others, or you can implement your own using TcpClient.

**Reply by mac-ro-ni** [reply]
I'm using SuperSimpleTCP Server to communicate with the other TCP clients but i'm confused about using the TCP received data as response for WatsonWebServer

**Reply by jchristn** [reply]
The received data should be a byte array that is given to you in an event.  You can use the ```HttpResponse.Send(byte[])``` method to send it.  Is it not working?

**Reply by mac-ro-ni** [reply]
How to access data received from tcp client inside a HTTPRoute?

**Reply by jchristn** [reply]
I see two approaches.  One, with SuperSimpleTcp, is to maintain a dictionary, key-value store, or table that is accessible to both the TCP client and the webserver.  When data is received by the TCP client, put it in that repository, and have the webserver continually monitor for new data.  The other is to use a library that provides explicit access to the underlying TCP socket, such as CavemanTcp.  That library will allow you to directly read from the socket from within your route.
